vcd/resource_vcd_independent_disk.go

```diff
@@ -173,33 +205,191 @@ func resourceVcdIndependentDiskCreate(d *schema.ResourceData, meta interface{})
+		if err != nil {
+			return diag.Errorf("error waiting to finish updating of independent disk: %s", err)
+		}
+
+		diagErr = attachBackVms(sliceOfVmsHrefs, vcdClient, disk, diskDetailsForReAttach)
+		if diagErr != nil {
+			return diagErr
+		}
+
+	}
+	return resourceVcdIndependentDiskRead(ctx, d, meta)
+}
+
+func lockVms(sliceOfVmsHrefs []string) {
+	for _, vmHref := range sliceOfVmsHrefs {
+		key := fmt.Sprintf("independentDisLock:%s", vmHref)
+		vcdMutexKV.kvLock(key)
+	}
+}
+
+func unlockVms(sliceOfVmsHrefs []string) {
+	for _, vmHref := range sliceOfVmsHrefs {
+		key := fmt.Sprintf("independentDisLock:%s", vmHref)
+		vcdMutexKV.kvUnlock(key)
+	}
+}
+
+func detachVms(vcdClient *VCDClient, disk *govcd.Disk, sliceOfVmsHrefs []string) (map[string]types.DiskSettings, diag.Diagnostics) {
+	diskDetailsForReAttach := make(map[string]types.DiskSettings)
+	var vms []*govcd.VM
+	for _, vmHref := range sliceOfVmsHrefs {
+		vm, err := vcdClient.Client.GetVMByHref(vmHref)
+		if err != nil {
+			return nil, diag.Errorf("error resourceVcdIndependentDiskUpdate error fetching attached VM: %s", err)
+		}
+		vmStatus := types.VAppStatuses[vm.VM.Status]
+		if vmStatus != "POWERED_OFF" && busTypesFromValues[disk.Disk.BusType] == "IDE" {
+			return nil, diag.Errorf("error resourceVcdIndependentDiskUpdate can not detach disks type `IDE` for VM `%s` which is not powered off", vm.VM.Name)
+		}
+		vms = append(vms, vm)
+	}
+
+	for _, vm := range vms {
+		isFoundDiskMatch := false
+		if vm.VM != nil && vm.VM.VmSpecSection != nil && vm.VM.VmSpecSection.DiskSection != nil && vm.VM.VmSpecSection.DiskSection.DiskSettings != nil {
+			for _, diskSettings := range vm.VM.VmSpecSection.DiskSection.DiskSettings {
+				if diskSettings.Disk != nil && diskSettings.Disk.HREF == disk.Disk.HREF {
+					diskDetailsForReAttach[vm.VM.HREF] = *diskSettings
+					isFoundDiskMatch = true
+				}
+			}
+		} else {
+			return nil, diag.Errorf("error resourceVcdIndependentDiskUpdate unexpected return from API, missing VmSpecSection or subtype")
+		}
+
+		if !isFoundDiskMatch {
+			return nil, diag.Errorf("error resourceVcdIndependentDiskUpdate couldn't match Disk with VM disk")
+		}
+		detachParams := &types.DiskAttachOrDetachParams{Disk: &types.Reference{HREF: disk.Disk.HREF}}
+
+		task, err := vm.DetachDisk(detachParams)
+		if err != nil {
+			return nil, diag.Errorf("error resourceVcdIndependentDiskUpdate error detaching independent disk `%s` to vm %s", disk.Disk.Name, err)
+		}
+		err = task.WaitTaskCompletion()
+		if err != nil {
+			return nil, diag.Errorf("error resourceVcdIndependentDiskUpdate error waiting for task to complete detaching independent disk `%s` to vm %s", disk.Disk.Name, err)
+		}
+	}
+	return diskDetailsForReAttach, nil
+}
+
+// attachBackVms reattaches independent disks back to VMs
+func attachBackVms(sliceOfVmsHrefs []string, vcdClient *VCDClient, disk *govcd.Disk, diskDetailsForReAttach map[string]types.DiskSettings) diag.Diagnostics {
```

Comment:
Just another nitpick :D
IMHO we could have same parameters order as `detachVms`
```suggestion
func attachBackVms(vcdClient *VCDClient, disk *govcd.Disk, sliceOfVmsHrefs []string, diskDetailsForReAttach map[string]types.DiskSettings) diag.Diagnostics {
```

Reply:
changed